common/src/main/kotlin/org/valkyrienskies/mod/common/assembly/AssemblyUtil.kt

```diff
@@ -0,0 +1,147 @@
+package org.valkyrienskies.mod.common.assembly
+
+import com.mojang.math.Vector3d
+import net.minecraft.core.BlockPos
+import net.minecraft.core.Direction
+import net.minecraft.nbt.CompoundTag
+import net.minecraft.util.Mth
+import net.minecraft.world.level.Level
+import net.minecraft.world.level.block.Block
+import net.minecraft.world.level.block.Blocks
+import net.minecraft.world.level.block.state.BlockState
+import net.minecraft.world.level.chunk.LevelChunk
+import net.minecraft.world.ticks.ScheduledTick
+import org.joml.Vector3i
+
+private val AIR = Blocks.AIR.defaultBlockState()
+object AssemblyUtil {
+
+    fun setBlock(level: Level, pos: BlockPos, state: BlockState?) {
+        val chunk = level.getChunk(pos) as LevelChunk
+        val section = chunk.getSection(chunk.getSectionIndex(pos.y))
+        val oldState = level.getBlockState(pos)
+        section.setBlockState(pos.x and 15, pos.y and 15, pos.z and 15, state)
+        ShipAssembler.triggerBlockChange(level, pos, oldState, state)
+    }
+
+    fun removeBlock(level: Level, pos: BlockPos) {
+        level.removeBlockEntity(pos)
+        setBlock(level, pos, Blocks.AIR.defaultBlockState())
+    }
+
+    fun copyBlock(level: Level, from: BlockPos?, to: BlockPos) {
+        val state = level.getBlockState(from)
+        val blockentity = level.getBlockEntity(from)
+        setBlock(level, to, state)
+
+        // Transfer pending schedule-ticks
+        if (level.blockTicks.hasScheduledTick(from, state.block)) {
+            level.blockTicks.schedule(ScheduledTick<Block?>(state.block, to, 0, 0))
+        }
+
+        // Transfer block-entity data
+        if (state.hasBlockEntity() && blockentity != null) {
+            val data: CompoundTag = blockentity.saveWithId()
+            level.setBlockEntity(blockentity)
+            val newBlockentity = level.getBlockEntity(to)
+            newBlockentity?.load(data)
+        }
+    }
+
+    fun updateBlock(level: Level, fromPos: BlockPos, toPos: BlockPos, toState: BlockState) {
+
+        // 75 = flag 1 (block update) & flag 2 (send to clients) + flag 8 (force rerenders)
```

Comment:
Why does it say 75 =, when its 11 and also flag 8 has a + instead of a &

Reply:
ewou, this PR is outdated. Tomato added the assembly code while they were adding vs-api to 1.18